Support logging exception stack traces when the last argument is Throwable

raccoonback opened the pull request:
## Motivation
Hello,

Reactor Core supports SLF4J, Console Logger, and JDK Logger.
In SLF4J implementations (e.g., Logback), when the last argument is of type Throwable, the exception stack trace is automatically logged, as shown below.
(ref. [SLF4J FAQ](https://www.slf4j.org/faq.html#paramException))
(ref: [Logback implementation](https://github.com/qos-ch/logback/blob/1da2f171dc108a2576eb33918198db7f755b1224/logback-classic/src/main/java/ch/qos/logback/classic/spi/LoggingEvent.java#L140-L142))

```java
log.info("An error occurred: {}", "details", new Exception("Sample exception"));
```

However, the Console Logger and JDK Logger supported by Reactor Core do not provide this functionality, and users must manually check for `Throwable` and handle it accordingly in their code.

```java
	@Test
	void useVerboseConsoleLoggers() throws Exception {
		try {
			Loggers.useVerboseConsoleLoggers();
			Logger l = Loggers.getLogger("test");

			l.debug("Processing data for user: {}", "user123", new Exception("Sample exception"));
		}
		finally {
			Loggers.resetLoggerFactory();
		        ReactorLauncherSessionListener.resetLoggersFactory();
		}
	}
```

```
[DEBUG] (Test worker) Processing data for user: user123
```

This results in unnecessary branching in user code and decreases consistency across logging mechanisms.

Therefore, in this change, we have modified the `ConsoleLogger` and `JdkLogger` to detect if the last argument is of type `Throwable` and automatically log the exception stack trace.

## Description

This PR enhances the logging functionality in Reactor Core to check if the last argument in logger methods is of type `Throwable`. If a `Throwable` is detected, the exception stack trace is included in the log output.

## Key Changes
1. Added logic to inspect the last argument in logger method calls.
2. Automatically handles `Throwable` instances for improved debugging.
3. Ensures compatibility with existing SLF4J-style message formatting for consistency across logging frameworks.

## Example Usage
```java
log.info("Processing data for user: {}", "user123", new Exception("Sample exception"));
```

### Before
```java
INFO: Processing data for user: user123
```
### After
```java
INFO: Processing data for user: user123
java.lang.Exception: Sample exception
    at ...

```

## Benefits
- Aligns Reactor's logging capabilities with familiar SLF4J-style formatting.
- Simplifies exception handling for developers using JDK or Console Loggers.
- Improves traceability and debugging in complex systems.

## Testing
```java

	@Test
	void useVerboseConsoleLoggers() throws Exception {
		try {
			Loggers.useVerboseConsoleLoggers();
			Logger l = Loggers.getLogger("test");

			l.debug("Processing data for user: {}", "user123", new Exception("Sample exception"));
		}
		finally {
			Loggers.resetLoggerFactory();
		        ReactorLauncherSessionListener.resetLoggersFactory();
		}
	}

	@Test
	void useJdkLoggers() throws Exception {
		try {
			Loggers.useJdkLoggers();
			Logger l = Loggers.getLogger("test");

			l.debug("Processing data for user: {}", "user123", new Exception("Sample exception"));
		}
		finally {
			Loggers.resetLoggerFactory();
		        ReactorLauncherSessionListener.resetLoggersFactory();
		}
	}
```

raccoonback commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -454,6 +455,24 @@ final String format(@Nullable String from, @Nullable Object... arguments){
+			logger.log(level, format(format, arguments));
+		}
+
+		private boolean isLastElementThrowable(Object... arguments) {
+			int length = arguments.length;
+			return length > 0 && arguments[length - 1] instanceof Throwable;
+		}
```
Checks whether the last argements are of type `Throwable`.

raccoonback commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -454,6 +455,24 @@ final String format(@Nullable String from, @Nullable Object... arguments){
 			return null;
 		}
+
+		private void logWithOptionalThrowable(Level level, String format, Object... arguments) {
+			if(isLastElementThrowable(arguments)) {
+				int lastIndex = arguments.length - 1;
+				Object[] args = Arrays.copyOfRange(arguments, 0, lastIndex);
+				Throwable t = (Throwable) arguments[lastIndex];
+
+				logger.log(level, format(format, args), t);
+				return;
+			}
+
+			logger.log(level, format(format, arguments));
+		}
```
In `JdkLogger`, if the last arguments are `Throwable` type, stacktrace is also logged.

chemicL replied:
I'd love to see a test for this.

raccoonback commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -507,6 +526,39 @@ final String format(@Nullable String from, @Nullable Object... arguments){
+			this.err.format("[%s] (%s) %s\n", level.toUpperCase(), Thread.currentThread().getName(), format(format, arguments));
+		}
+
+		private boolean isLastElementThrowable(Object... arguments) {
+			int length = arguments.length;
+			return length > 0 && arguments[length - 1] instanceof Throwable;
+		}
```
Checks whether the last argements are of type `Throwable`.

raccoonback commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -507,6 +526,39 @@ final String format(@Nullable String from, @Nullable Object... arguments){
 			return null;
 		}
 
+		private synchronized void logWithOptionalThrowable(String level, String format, Object... arguments) {
+			if(isLastElementThrowable(arguments)) {
+				int lastIndex = arguments.length - 1;
+				Object[] args = Arrays.copyOfRange(arguments, 0, lastIndex);
+				Throwable t = (Throwable) arguments[lastIndex];
+
+				this.log.format("[%s] (%s) %s\n", level.toUpperCase(), Thread.currentThread().getName(), format(format, args));
+				t.printStackTrace(this.log);
+				return;
+			}
+
+			this.log.format("[%s] (%s) %s\n", level.toUpperCase(), Thread.currentThread().getName(), format(format, arguments));
+		}
```
In `ConsoleLogger`, if the last argument is of type `Throwable`, it also prints a stack trace.
(only use for TRACE, DEBUG, INFO level)

chemicL replied:
Can we have tests to validate this?

raccoonback replied:
I have added test code!

raccoonback commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -507,6 +526,39 @@ final String format(@Nullable String from, @Nullable Object... arguments){
+			this.log.format("[%s] (%s) %s\n", level.toUpperCase(), Thread.currentThread().getName(), format(format, arguments));
+		}
+
+		private synchronized void logErrorWithOptionalThrowable(String level, String format, Object... arguments) {
+			if(isLastElementThrowable(arguments)) {
+				int lastIndex = arguments.length - 1;
+				Object[] args = Arrays.copyOfRange(arguments, 0, lastIndex);
+				Throwable t = (Throwable) arguments[lastIndex];
+
+				this.err.format("[%s] (%s) %s\n", level.toUpperCase(), Thread.currentThread().getName(), format(format, args));
+				t.printStackTrace(this.err);
+				return;
+			}
+
+			this.err.format("[%s] (%s) %s\n", level.toUpperCase(), Thread.currentThread().getName(), format(format, arguments));
+		}
```
In `ConsoleLogger`, if the last arguments are `Throwable` type, it also prints a stack trace with the `err` PrintStream.
(only use for WARN, ERROR level)

raccoonback commented:
@chemicL 
Hello!
Could you please take a look at this PR?
The review seems to be delayed, so I would greatly appreciate your confirmation.
Thank you!

chemicL commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -507,6 +526,39 @@ final String format(@Nullable String from, @Nullable Object... arguments){
 			return null;
 		}
 
+		private synchronized void logWithOptionalThrowable(String level, String format, Object... arguments) {
```
Same comments as with the JdkLogger.

chemicL commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -507,6 +526,39 @@ final String format(@Nullable String from, @Nullable Object... arguments){
+			this.log.format("[%s] (%s) %s\n", level.toUpperCase(), Thread.currentThread().getName(), format(format, arguments));
+		}
+
+		private synchronized void logErrorWithOptionalThrowable(String level, String format, Object... arguments) {
```
This method is almost exactly the same as the one above, can you please unify them with a `PrintStream` as an argument?

raccoonback replied:
I have merged the methods into one, and modified it to accept a `PrintStream` as a parameter.

chemicL commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -573,17 +625,17 @@ public boolean isInfoEnabled() {
-		public synchronized void info(String format, Object... arguments) {
-			this.log.format("[ INFO] (%s) %s\n", Thread.currentThread().getName(), format(format, arguments));
+		public void info(String format, Object... arguments) {
+			logWithOptionalThrowable("INFO", format, arguments);
```
I am guessing the space was to align INFO under DEBUG and such in a single column so the space was a filler. With the spaces all of (TRACE, DEBUG, INFO, WARN, ERROR) take up the same amount of horizontal space. Please roll back.

raccoonback replied:
Thanks for detailed explanation!
I have reverted the code.

chemicL reviewed (changes requested):
Thank you for the proposal. I've been occupied with other priority subjects and now found the time to review the PR. I think it is a useful utility. I left some comments, please address them if you'd like to integrate this change. Most importantly, I look forward to tests validating the change has some meaningful effect.

raccoonback commented:
@chemicL 
First of all, thanks for your review!
I have incorporated all of your suggestions.
Please let me know if there's anything else that needs adjustment!

raccoonback commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -454,6 +455,24 @@ final String format(@Nullable String from, @Nullable Object... arguments){
+				Object[] args = Arrays.copyOfRange(arguments, 0, lastIndex);
+				Throwable t = (Throwable) arguments[lastIndex];
+
+				logger.log(level, format(format, args), t);
```
I added the `skipLast` parameter to the `format(...)` method. Since `Varargs` must be placed at the end, I added `skipLast` as a **middle** parameter. 
Additionally, I overloaded the existing method.

raccoonback commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -494,19 +521,57 @@ public String getName() {
 			return null;
 		}
 
+		private synchronized void logWithPotentialThrowable(PrintStream logger, String level, String format, Object... arguments) {
```
add a `PrintStream` parameter.

chemicL commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -442,18 +442,45 @@ public void error(String msg, Throwable t) {
 		@Nullable
-		final String format(@Nullable String from, @Nullable Object... arguments){
-			if(from != null) {
+		String format(@Nullable String from, @Nullable Object... arguments) {
```
This one can also be private.

raccoonback replied:
Changed to `private` accessor.

chemicL commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -442,18 +442,45 @@ public void error(String msg, Throwable t) {
-					for (Object argument : arguments) {
-						computed = computed.replaceFirst("\\{\\}", Matcher.quoteReplacement(String.valueOf(argument)));
+					int lastIndex = arguments.length;
+					if (skipLast) --lastIndex;
```
Please make sure you use braces around the instructions following an if condition.

raccoonback replied:
Reflect it!

chemicL commented on reactor-core/src/main/java/reactor/util/Loggers.java:
```diff
@@ -442,18 +442,45 @@ public void error(String msg, Throwable t) {
+		}
+
+		@Nullable
+		private String format(@Nullable String from, boolean skipLast, @Nullable Object... arguments) {
```
> I added the skipLast parameter to the format(...) method. Since Varargs must be placed at the end, I added skipLast as a middle parameter.
Additionally, I overloaded the existing method.

Would it work if instead of `Object...` you used `Object[]` ? Since this is an internal API it should be totally fine with the usages here. Then the order of arguments will be more appropriate IMO.

raccoonback replied:
I have changed it to use `Object[]` instead.

chemicL reviewed (changes requested):
Thanks for the follow up. I added a few more comments, please apply them also to similar areas where appropriate. Please check the build results, they are failing with NPE. Before the next round please make sure the build is passing on your machine.

raccoonback commented:
@chemicL 
Hello.

Thank you for your detailed review!
I have addressed all the comments and also fixed the issue causing the NPE.

chemicL reviewed (changes requested):
Thank you for the follow up. The production code looks good. Regarding the tests, please refrain from using mocks when more end-to-end tests are possible. The existing tests allow validating the output. Check the use of `ConsoleLoggerTest#outContent` and `JdkLoggerTest#nullFormatIsAcceptedByUnderlyingLogger`. Also, please revert unrelated formatting changes so that the PR consists only of changes relevant to the problem that's being solved. Thanks.

raccoonback commented:
@chemicL 

I have rolled back the unrelated formatting changes.

Additionally, since ConsoleLoggerTest already covers all cases for ConsoleLogger, I found the newly added test to be redundant and have therefore removed it.

chemicL commented:
> I have rolled back the unrelated formatting changes.
> 
> Additionally, since ConsoleLoggerTest already covers all cases for ConsoleLogger, I found the newly added test to be redundant and have therefore removed it.

I'm not sure I understand. Since the functionality of `ConsoleLogger` has been enhanced, there should be a need for more tests. In `JdkLoggerTest` there is no validation of the stacktrace. Please thoroughly review the changes and expectations. Also, make sure the test methods are descriptive as the existing methods are and capture the circumstance and expectation.

raccoonback commented:
@chemicL 

I misunderstood;;
As you mentioned, I have added a test for the case where the last argument of varargs is a `Throwable`.
Additionally, I modified `JdkLoggerTest` to validate the stacktrace and extracted the duplicate code for creating JdkLogger into the `getJdkLogger()` method.

chemicL commented:
@raccoonback thank you for the contribution! 🚀
